### ISSUE TITLE
[73] Why Do Better Loss Functions Lead to Less Transferable Features?

### ISSUE THREAD

**Issue by dhkim0225** [dhkim0225/1day_1paper#102]
ImageNet 성능이 좋다고 transferability가 좋은게 아님. 
9개의 loss 에 따라 transferability 가 어떻게 달라지는지 분석해 보자.

[paper](https://arxiv.org/pdf/2010.16402.pdf)
[sungchul.kim review KR (notion)](https://www.notion.so/Why-Do-Better-Loss-Functions-Lead-to-Less-Transferable-Features-75d77d79294241c1b3f043740b726858)

# INTRO
1. objective는 network layer의 representation이 output과 얼마나 가까운지에 영향을 미침. 
2. CKA (Centered Kernel Alignment)로 확인해보니 마지막 레이어 몇 개의 representation은 loss 를 뭐로 쓰냐에 따라 완전히 달랐고 초기 레이어에서의 feature들은 어떤 loss를 사용하든 유사했음.

3. original task에서 softmax ce보다 좋은 성능을 보인 objective들은 끝에서 두번째 레이어에서 다른 클래스들의 representation사이에 강한 separation이 일어남. original task에서 좋은 성능을 보이고 downstream task에서 안좋은 성능을 보이는 `within-class variability`가 붕괴됨을 보임

# 9 Losses
## Preliminary
<img src="https://render.githubusercontent.com/render/math?math=\LARGE z \in \mathbb{R}^K"> : logit

<img src="https://render.githubusercontent.com/render/math?math=\LARGE t \in \{0, 1\}^K"> : one-hot vector target

<img src="https://render.githubusercontent.com/render/math?math=\LARGE x \in \mathbb{R}^M"> : 끝 layer vector activation

<img src="https://render.githubusercontent.com/render/math?math=\LARGE z = Wx %2B b (W \in \mathbb{R}^{K \times M})">

## Softmax cross-entropy
![image](https://user-images.githubusercontent.com/16400591/155066990-c5521e10-4243-407f-9cad-ee91ae5f96f4.png)

## Label smoothing
![image](https://user-images.githubusercontent.com/16400591/155067008-cc946bc5-200a-445f-9da4-dfc6e3c8b389.png)

## Dropout on penultimate layer
![image](https://user-images.githubusercontent.com/16400591/155067024-00303e19-48c0-46fa-bad9-8df5c3fc28b8.png)

## Extra final layer L^2 regularization
final layer 에 큰 l2 reg 걸면 효과가 있음
![image](https://user-images.githubusercontent.com/16400591/155067049-235ff1c0-8d80-424e-a9a4-13f6a7b1b1cd.png)

## Logit penalty
logit 의 l2 norm 제약. (dropout 과 유사한 성능 보임)
https://openreview.net/forum?id=d-XzF81Wg1
![image](https://user-images.githubusercontent.com/16400591/155067075-1719bf76-cf68-4a6c-bd08-2b3e0c3a4ea5.png)

## Logit normalization
![image](https://user-images.githubusercontent.com/16400591/155067107-7dfb2781-be70-429a-a9ca-d21ff8ddca9a.png)

## Cosine softmax
arcface 등에서 사용하는 loss 형태
사실 cross_entropy term 에서 z_t 만 cosine similarity 로 바꿔치기 한 것과 다르지 않음.
![image](https://user-images.githubusercontent.com/16400591/155067136-7fa27c7a-f4b3-4bf8-9a23-094ddf9998b5.png)

## Sigmoid cross-entropy
각 dimension 별로 BCE 로 학습해서 최종 output 은 score 제일 높은 애 고르기
![image](https://user-images.githubusercontent.com/16400591/155067163-b7f9f444-9145-4787-b57a-4c20a1bea402.png)

## Squared error
![image](https://user-images.githubusercontent.com/16400591/155067178-f082b51e-72f0-4b04-89e7-1102a6bbd537.png)

# Result
## Linear Probing, KNN
![image](https://user-images.githubusercontent.com/16400591/155074848-057241e5-6385-4f8e-bf7e-9b72e806e8d5.png)
linear probing 이나 KNN 으로 transfer 해봤음.
성능 좋은게 transferability 가 높지는 않더라.
![image](https://user-images.githubusercontent.com/16400591/155075231-09b9e825-ccee-4c57-a6e3-2d0bb1b782e2.png)

## Fine-Tuning
![image](https://user-images.githubusercontent.com/16400591/155075151-f05b0468-4f43-4288-af8a-dc411618c87a.png)

### CKA_Linear
![image](https://user-images.githubusercontent.com/16400591/155075441-5b434989-e881-455e-87f8-e0e13f52c783.png)

마지막 쪽으로 갈수록 loss 별 representation 이 다르더라.
![image](https://user-images.githubusercontent.com/16400591/155075385-65cbbf29-9c8c-4865-bdf3-05d9f50b79c7.png)
![image](https://user-images.githubusercontent.com/16400591/155075604-3bcad51e-8960-45c1-b798-8d8c2ae42ac2.png)

### Activation Outs
layer 별 nonzero activation 도 찍어 봤는데,
imagenet 에서는 확 성능 좋았던 애들이 transfer 잘 안되던,
`logit norm` `cosine softmax`  애들이 activation 활성도도 떨어지더라.
~`sigmoid` 는 안 그런데 흠..~
![image](https://user-images.githubusercontent.com/16400591/155075836-de5db0d2-3416-4965-aa7c-7f7476ddd872.png)

### class separation
regularizer 를 쓸 수록 separation 이 증가하더라.
![image](https://user-images.githubusercontent.com/16400591/155076783-f85155f4-da23-4234-abef-8e7316b351a5.png)

즉, class 에 대해서 특화된 분류기가 만들어지는 느낌?

cosine softmax 에서 temperature 를 올려보니까 class separation 수치도 올라가던데,
그러면 transferability 가 떨어지더라.

![image](https://user-images.githubusercontent.com/16400591/155076989-5df514e5-aa19-45ae-aa11-bb51a5164fab.png)

### Pets dataset 은 왜 잘되냐?
Oxford-IIIT Pets의 클래스가 37개인데 그 중 25개가 ImageNet의 클래스더라.
![image](https://user-images.githubusercontent.com/16400591/155077532-a1e63d76-de77-4bad-a33e-dd2b69a52296.png)

### 어떤 loss 들 끼리 특성이 비슷할까?
CKA 로 그래프 뽑아내 봤다.
![image](https://user-images.githubusercontent.com/16400591/155077418-888a9e1c-0e64-4928-ad41-e508b8128cb6.png)
